server/worker/apple_mdm.go

```diff
@@ -112,30 +127,126 @@ func (a *AppleMDM) runPostDEPEnrollment(ctx context.Context, args appleMDMArgs)
 
 		if ssoEnabled {
 			a.Log.Log("info", "setting username and fullname", "host_uuid", args.HostUUID)
+			cmdUUID := uuid.New().String()
 			if err := a.Commander.AccountConfiguration(
 				ctx,
 				[]string{args.HostUUID},
-				uuid.New().String(),
+				cmdUUID,
 				acct.Fullname,
 				acct.Username,
 			); err != nil {
 				return ctxerr.Wrap(ctx, err, "sending AccountConfiguration command")
 			}
+			awaitCmdUUIDs = append(awaitCmdUUIDs, cmdUUID)
+		}
+	}
+
+	var manualRelease bool
+	if args.TeamID == nil {
+		ac, err := a.Datastore.AppConfig(ctx)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "get AppConfig to read enable_release_device_manually")
+		}
+		manualRelease = ac.MDM.MacOSSetup.EnableReleaseDeviceManually.Value
+	} else {
+		tm, err := a.Datastore.Team(ctx, *args.TeamID)
+		if err != nil {
+			return ctxerr.Wrap(ctx, err, "get Team to read enable_release_device_manually")
+		}
+		manualRelease = tm.Config.MDM.MacOSSetup.EnableReleaseDeviceManually.Value
+	}
+
+	if !manualRelease {
+		// send all command uuids for the commands sent here during post-DEP
+		// enrollment and enqueue a job to look for the status of those commands to
+		// be final and same for MDM profiles of that host; it means the DEP
+		// enrollment process is done and the device can be released.
+		if err := QueueAppleMDMJob(ctx, a.Datastore, a.Log, AppleMDMPostDEPReleaseDeviceTask,
+			args.HostUUID, args.TeamID, args.EnrollReference, awaitCmdUUIDs...); err != nil {
+			return ctxerr.Wrap(ctx, err, "queue Apple Post-DEP release device job")
+		}
+	}
+
+	return nil
+}
+
+func (a *AppleMDM) runPostDEPReleaseDevice(ctx context.Context, args appleMDMArgs) error {
+	// Edge cases:
+	//   - if the device goes offline for a long time, should we go ahead and
+	//   release after a while?
+	//   - if some commands/profiles failed (a final state), should we go ahead
+	//   and release?
+	//   - if the device keeps moving team, or profiles keep being added/removed
+	//   from its team, it's possible that its profiles will never settle and
+	//   always have pending statuses. Same as going offline, should we release
+	//   after a while?
+	//
+	// We opted "yes" to all those, and we want to release after a few minutes,
+	// not hours, so we'll allow only a couple retries.
+
+	if retryNum, _ := ctx.Value(retryNumberCtxKey).(int); retryNum > 2 {
+		// give up and release the device
+		a.Log.Log("info", "releasing device after too many attempts", "host_uuid", args.HostUUID)
+		if err := a.Commander.DeviceConfigured(ctx, args.HostUUID, uuid.NewString()); err != nil {
+			return ctxerr.Wrap(ctx, err, "failed to enqueue DeviceConfigured command")
+		}
+		return nil
+	}
+
+	for _, cmdUUID := range args.EnrollmentCommands {
+		res, err := a.Datastore.GetMDMAppleCommandResults(ctx, cmdUUID)
```

Comment:
one _possible_ improvement once we start labeling commands as "fleet initiated" could be to add that filter here, this way we're not waiting on any commands that might be issued by the IT admin.

Reply:
(and those are all commands targeting a single host)

Reply:
you're right! sorry for missing that!